src/profile-logic/process-profile.js

```diff
@@ -1052,56 +1056,195 @@ export function processProfile(
+      })
+      .filter(thread => thread !== null),
   });
+
   return JSON.stringify(newProfile);
 }
 
+/**
+ * Take a thread with PII that user wants to be removed and remove the thread
+ * data depending on that PII status. Look at the `PII` type definition if you
+ * want to learn what kind of information we are removing.
+ */
+function removeThreadPII(
+  thread: Thread,
+  threadIndex: number,
+  stringArray: string[],
+  PIIToBeRemoved: PII
+): Thread | null {
+  if (
+    PIIToBeRemoved.hiddenThreads !== null &&
+    PIIToBeRemoved.hiddenThreads.includes(threadIndex)
+  ) {
+    // If this is a hidden thread, remove the thread immediately.
+    // This will not remove the thread entry from the `threads` array right now
+    // and just replace it with a `null` value. We filter out the null values
+    // inside `serializeProfile` function.
+    return null;
+  }
+
+  // We iterate all the markers and remove/change data depending on the PII
+  // status.
+  const markersToDelete = [];
+  if (
+    PIIToBeRemoved.networkUrls ||
+    PIIToBeRemoved.screenshots === true ||
+    (Array.isArray(PIIToBeRemoved.screenshots) &&
+      PIIToBeRemoved.screenshots.length > 0)
+  ) {
+    for (let i = 0; i < thread.markers.length; i++) {
+      const currentMarker = thread.markers.data[i];
+
+      // Remove the all network URLs if user wants to remove them.
+      if (
+        PIIToBeRemoved.networkUrls &&
+        currentMarker &&
+        currentMarker.type &&
+        currentMarker.type === 'Network'
+      ) {
+        // Remove the URI fields from marker payload.
+        currentMarker.URI = '';
+        currentMarker.RedirectURI = '';
+        // Strip the URL from the marker name
+        const stringIndex = thread.markers.name[i];
+        stringArray[stringIndex] = stringArray[stringIndex].replace(/:.*/, '');
+      }
+
+      // If PIIToBeRemoved.screenshots is true, this means that we need to
+      // remove all screenshots, if it's an array of thread IDs, that means we
+      // should only delete screenshots from the given threads.
+      if (
+        (PIIToBeRemoved.screenshots === true ||
+          (Array.isArray(PIIToBeRemoved.screenshots) &&
+            PIIToBeRemoved.screenshots.includes(threadIndex))) &&
+        currentMarker &&
+        currentMarker.type &&
+        currentMarker.type === 'CompositorScreenshot'
+      ) {
+        const urlIndex = currentMarker.url;
+        stringArray[urlIndex] = '';
+        markersToDelete.push(i);
+      }
+    }
+  }
+
+  // After iterating (or not iterating at all) the markers, if we have some
+  // markers we want to delete or user wants to delete the full time range,
+  // reconstruct the marker table and samples table without unwanted information.
+  if (markersToDelete.length > 0 || PIIToBeRemoved.fullTimeRange !== null) {
+    const oldMarkers = thread.markers;
+    const newMarkerTable = getEmptyRawMarkerTable();
+    const addCurrentMarker = (index: number) => {
```

Comment:
nit: 2 things here: this is not adding the "current" marker, and this is not always adding a marker.
So  maybe rename to `possiblyAddMarkerIndex` or `addMarkerIndexIfIncluded` ?